### ISSUE TITLE
Feature Request: Include Depthwise Convolution in graph_transforms fold_batch_norms

### ISSUE THREAD

**Issue by christopherhelf** [tensorflow/tensorflow#11370]
Hi!

Currently, the graph_transforms tool includes only `Conv2D` and `MatMul` ops when folding batch normalization scaling/multiplication into its weights, as in `fold_batch_norms.cc`. Google's Mobilenet example uses depthwise convolution extensively, so it would be nice to include this feature for the `DepthwiseConv2dNative` operation. The problem here is that weights are ordered differently for this operation and contain the `channel_muliplier` which would need to be checked when trying to bake subsequent multiplications.

**Comment by petewarden** [comment]
This may be a bit tricky to implement, since as mentioned the depthwise convolution does its math a bit differently from the standard version. I'd like to see this added too, but I'm not sure of a timeframe on when we'll get to it, so adding to contributions welcome as well.

**Comment by christopherhelf** [comment]
Would this be the right approach in general?

So `Conv2D` has its filters as
    [filter_height, filter_width, in_channels, out_channels]

while `DepthwiseConv2dNative` has 
    [filter_height, filter_width, in_channels, channel_multiplier]

So in the simple case of `channel_multiplier=1`, we have `out_channels=in_channels`, and `mul_values.shape().dim_size(0) == in_channels` and something like

    auto weights_mapped = weights.tensor<float, 4>();
    auto scaled_weights_mapped = scaled_weights.tensor<float, 4>();

    for (int64 cm = 0; cm < weights.dim_size(3); ++cm) {
      for (int64 ic = 0; ic < weights.dim_size(2); ++ic) {
        for (int64 w = 0; w < weights.dim_size(1); ++w) {
          for (int64 h = 0; h < weights.dim_size(0); ++h) {
            scaled_weights_mapped(b,w,ic,cm) = weights_mapped(b,w,ic,cm) * mul_values.flat<float>()(ic);
          }
        }
      }
    }

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by hdeeb** [comment]
We also need this in our project. We will really appreciate it if you give timeline when it will be added at all. We need to know if it is not planned for the near future. In this case we need to develop it in our code.
I'll appreciate your quick reply.
Thanks

**Comment by chuan92** [comment]
This script may help you. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/quantize/python/fold_batch_norms.py . In deed, if you run tf 1.9, grappler always optimize batch normalization when possible.

**Comment by wt-huang** [comment]
Closing as this is resolved

**Comment by frreiss** [comment]
@petewarden I see this issue was closed without a fix being added to the Graph Transform Tool. Would you be interested in a PR that adds depthwise convolution support to the C++ implementations of the `fold_batch_norms` and `fold_old_batch_norms` rewrites?

I've already implemented depthwise convolution support in my own Python version of `fold_*_batch_norms`; code is [here](https://github.com/CODAIT/graph_def_editor/blob/5695fdc9381ee0f935d1ece7547e5bf111cae25e/graph_def_editor/rewrite.py#L194).
